File: workspace/oscillodemo/oscillo.c
/*
Copyright (c) 2020, Chubu University and Firmlogics

All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

1. Redistributions of source code must retain the above copyright notice, this
   list of conditions and the following disclaimer.

2. Redistributions in binary form must reproduce the above copyright notice,
   this list of conditions and the following disclaimer in the documentation
   and/or other materials provided with the distribution.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/

/*
 * OscilloDSP: Virtual Oscilloscope Library
 *
 * Written by Atsushi Yokoyama, Firmlogics (contact@flogics.com)
 */


#undef DEBUG

#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <string.h>
#include <assert.h>
#include <math.h>
#include "config.h"
#include "common.h"
#include "oscillo.h"
#include "com.h"
#include "oscillodsp.pb.h"

#ifndef M_PI
#define M_PI    3.14159265358979323846264338327950288
#endif /* !defined(M_PI) */

enum {
    N_CHANNELS = pb_arraysize(WaveGroup, wave),
    N_WAVE_SAMPLES = pb_arraysize(Wave, samples),
};

static buffer_t buffer[N_CHANNELS][LEN_BUFFER];
static ConfigReply config_reply;
static dsp_channel_config_t dsp_ch_config[N_CHANNELS];
static int idx_write_buffer[N_CHANNELS];
static Configure config;


/**
 * Correct buffer index
 *
 * Refer https://www.lemoda.net/c/modulo-operator/
 */
#define MOD(a, b) ((((a) % (b)) + (b)) % (b))
static int correct_idx(int idx)
{
    return MOD(idx, LEN_BUFFER);
}


/**
 * Convert float value to buffer_t value
 *
 * @param ch corresponding channel ID
 * @param val is original float value
 */
static buffer_t f_to_bufval(int ch, float val, int *err)
{
    long tmp_val;
    long drange;
    long max_val;
    long min_val;
    float mul;
    float offset;
    ChannelConfig *cc;

    if (ch < 0 || ch >= N_CHANNELS) {
        *err = -1;
        return 0;
    }

    if (! dsp_ch_config[ch].enabled) {
        *err = -2;
        return 0;
    }

    cc = &config_reply.chconfig[ch];

    /*
     * Prepare constant values to convert val to integer which fits buffer_t.
     */
    drange = 1 << (config.resolution - 1);

    // corresponds to 0x7f...ff (in two's complement)
    max_val = drange - 1;
    // corresponds to 0x80...00 (ditto)
    min_val = - drange;
    // coefficient multiplying to the original value
    mul = 2.0f * drange / (cc->max - cc->min);
    // center of the signal range
    offset = (cc->max + cc->min) / 2.0f;

    /*
     * Convert val and check if the converted value doesn't overflow.
     */
    tmp_val = (long) ((val - offset) * mul);
    if (tmp_val > max_val)
        tmp_val = max_val;
    else if (tmp_val < min_val)
        tmp_val = min_val;

    *err = 0;
    return (buffer_t) tmp_val;
}


/**
 * Initialize oscillo library
 *
 * @param sample_rate is sampling rate [Hz]
 * @param default_timescale is length of frame width time [seconds]
 */
int oscillo_init(float sample_rate, float default_timescale)
{
    memset(buffer, 0, sizeof(buffer));
    memset(&config_reply, 0, sizeof(config_reply));
    memset(dsp_ch_config, 0, sizeof(dsp_ch_config));
    memset(idx_write_buffer, 0, sizeof(idx_write_buffer));

    config.resolution = 8;  // set because resolution '0' is not good default
    config.trigmode = TriggerMode_Auto;
    config.trigtype = TriggerType_RisingEdge;
    config.ch_trig = 0;
    config.triglevel = 0.0f;

    config_reply.samplerate = sample_rate;
    config_reply.default_timescale = default_timescale;
    config_reply.max_timescale = (float) LEN_BUFFER / sample_rate;

    return 0;
}


/**
 * Re-initialize oscillo library
 */
int oscillo_reinit(void)
{
    return oscillo_init(
            config_reply.samplerate, config_reply.default_timescale);
}


/**
 * Pass a sample value to oscillo library
 *
 * @param ch is channel ID starting at 0
 * @param val is passed sample value.  val _should_ be between -1 to 1.
 */
int oscillo_pass_one(int ch, sample_t val)
{
    int err;

    if (ch < 0 || ch >= N_CHANNELS)
        return -1;

    if (! dsp_ch_config[ch].enabled)
        return -2;

    buffer[ch][idx_write_buffer[ch]] = f_to_bufval(ch, val, &err);
    if (err)
        return err;

    /*
     * Update idx_write_buffer
     */
    idx_write_buffer[ch] = correct_idx(idx_write_buffer[ch] + 1);

    return 0;
}


/**
 * Process oscillo library
 *
 * @param val is passed sample value
 */
int oscillo_proc(void)
{
    return 0;
}


/**
 * Configure oscillo mode
 */
int oscillo_config(Configure arg_config, ConfigReply *arg_config_reply)
{
    if (arg_config.resolution > sizeof(buffer_t) * 8)
        return -1;

    if (arg_config.trigmode < TriggerMode_Auto)
        return -1;

    if (arg_config.trigmode > TriggerMode_Single)
        return -1;

    if (arg_config.trigtype < TriggerType_RisingEdge)
        return -1;

    if (arg_config.trigtype > TriggerType_FallingEdge)
        return -1;

    if (arg_config.ch_trig >= N_CHANNELS)
        return -1;

    if (! dsp_ch_config[arg_config.ch_trig].enabled)
        return -2;

    memcpy(&config, &arg_config, sizeof(Configure));

    /*
     * If specified timescale is 0.0f as a special value, we recover default
     * timescale.
     */
    if (arg_config.timescale == 0.0f)
        config.timescale = config_reply.default_timescale;

    memcpy(arg_config_reply, &config_reply, sizeof(ConfigReply));

    return 0;
}


/**
 * Configure oscillo channel
 *
 * @param name is channel name
 * @param unit is unit (V, A, sec, etc.)
 * @param min is expected minimum value of any samples
 * @param max is expected maximum value of any samples
 *
 * @return channel ID starting at 0
 */
int oscillo_config_ch(const char *name, const char *unit, float min, float max)
{
    int ch;
    ChannelConfig *cc;

    if (config_reply.chconfig_count >= N_CHANNELS)
        return -1;

    ch = config_reply.chconfig_count;
    cc = &config_reply.chconfig[ch];

    strncpy(cc->name, name, pb_arraysize(ChannelConfig, name) - 1);
    cc->name[pb_arraysize(ChannelConfig, name) - 1] = '\0';

    strncpy(cc->unit, unit, pb_arraysize(ChannelConfig, unit) - 1);
    cc->unit[pb_arraysize(ChannelConfig, unit) - 1] = '\0';

    cc->min = min;
    cc->max = max;

    dsp_ch_config[ch].enabled = true;

    config_reply.chconfig_count ++;

    return ch;
}


/**
 * Scale an index in wave samples, to index in buffer[][] according to
 * timescale and sampling rate
 *
 * @param idx is index in wave samples
 */
static int scale_index(int idx)
{
    float t = (float) idx / N_WAVE_SAMPLES * config.timescale;

    return (int) (t * config_reply.samplerate);
}


/**
 * Compare for triggering edge detection
 *
 * For rising edge, return true if v1 > v2.
 * Return false in the falling edge case.
 *
 * @param v1, v2
 */
static bool comp_trig_edge(buffer_t v1, buffer_t v2)
{
    if (config.trigtype == TriggerType_RisingEdge)
        return v1 > v2;
    else
        return v1 < v2;
}


/**
 * Determine tail sample position to send
 *
 * @return index in buffer
 */
static int get_pos(bool *triggered)
{
    int i;
    int ch;
    int err;
    int pos_buf;
    int tail_pos;
    int int_triglevel;
    int margin_x;
    int margin_y;
    buffer_t cur_val;
    buffer_t past_val;
    bool found;
    bool cond_met;

    ch = config.ch_trig;
    int_triglevel = f_to_bufval(ch, config.triglevel, &err);
    if (err)
        return err;

    found = false;

    cond_met = false;
    tail_pos = correct_idx(idx_write_buffer[ch] - 1);
    margin_x = (int) ((float) HIST_MARGIN_X * N_WAVE_SAMPLES);
    margin_y = (int) ((float) HIST_MARGIN_Y * (1 << config.resolution));

    for (i = 0; i < N_WAVE_SAMPLES; i ++) {
        pos_buf = correct_idx(tail_pos - scale_index(i + N_WAVE_SAMPLES / 2));

        assert(pos_buf >= 0);
        assert(pos_buf < LEN_BUFFER);

        cur_val = buffer[ch][pos_buf];
        if (comp_trig_edge(cur_val, int_triglevel))
            cond_met = true;

        past_val = buffer[ch][correct_idx(pos_buf - margin_x)];
        if (cond_met &&
                comp_trig_edge(cur_val - margin_y, past_val) &&
                comp_trig_edge(int_triglevel, cur_val)) {
            found = true;
            pos_buf = correct_idx(pos_buf + scale_index(N_WAVE_SAMPLES / 2));
            break;
        }
    }

    *triggered = found;

    if (! found) {
        if (config.trigmode == TriggerMode_Normal)
            pos_buf = -1;
        else
            pos_buf = correct_idx(idx_write_buffer[ch] - 1);
    }

    return pos_buf;
}


/**
 * Return a wave frame
 */
static int get_wave(int ch, int pos_buf, Wave *wave)
{
    int i;
    int idx;

    if (ch < 0 || ch >= N_CHANNELS)
        return -1;

    if (! dsp_ch_config[ch].enabled)
        return -2;

    wave->ch_id = ch;

    assert(pos_buf >= 0);
    assert(pos_buf <= LEN_BUFFER);

    for (i = 0; i < N_WAVE_SAMPLES; i ++) {
        idx = correct_idx(pos_buf - scale_index(i));

        assert(idx >= 0);
        assert(idx < LEN_BUFFER);

        /*
         * Copy samples in buffer[ch][] to wave.samples[]
         */
        wave->samples[N_WAVE_SAMPLES - 1 - i] = buffer[ch][idx];
    }

    return 0;
}


/**
 * Return group of wave frames
 */
int oscillo_get_waves(WaveGroup *waves)
{
    int ch;
    int pos_buf;
    bool triggered;

    pos_buf = get_pos(&triggered);

    if (pos_buf < 0) {  // Implies TriggerMode_Normal
        waves->triggered = false;
        waves->wave_count = 0;
        return 0;
    }

    /*
     * Only when triggered, send wave frames
     */
    for (ch = 0; ch < config_reply.chconfig_count; ch ++)
        get_wave(ch, pos_buf, &waves->wave[ch]);

    waves->triggered = triggered;
    waves->wave_count = config_reply.chconfig_count;

    return 0;
}


/*
 * Return a float random value
 *
 * @param amp is amplitude of the random value
 */
static float get_noise(float amp)
{
    return amp * (float) rand() / RAND_MAX - amp / 2;
}


/**
 * Return a demo (version 1) value
 */
float oscillo_get_demo1_value(bool enabled)
{
    static int deg = 0;
    float val;
    float noise;

    noise = get_noise(0.12);
    val = 0.5 + noise;
    if (enabled) {
        val += 3.0 * sinf((float) deg / 180 * M_PI);
        val += 0.3 * sinf((float) deg * 8 / 180 * M_PI);
    }

    deg ++;
    if (deg >= 360)
        deg = 0;

    return val;
}


/**
 * Return a demo (version 2) value
 */
float oscillo_get_demo2_value(bool enabled)
{
    static int deg = 0;
    float val;
    float noise;

    noise = get_noise(0.12);
    val = noise;
    if (enabled)
        val += 1.5 * (deg % 180 < 90 ? 1.0 : -1.0);

    deg ++;
    if (deg >= 360)
        deg = 0;

    return val;
}

File: hostapp/oscillodsp/oscillo.py
"""
Copyright (c) 2020-2021, Chubu University and Firmlogics

All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

1. Redistributions of source code must retain the above copyright notice, this
   list of conditions and the following disclaimer.

2. Redistributions in binary form must reproduce the above copyright notice,
   this list of conditions and the following disclaimer in the documentation
   and/or other materials provided with the distribution.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
"""

import asyncio
import logging
import math
import sys

import ipywidgets as widgets
import matplotlib.pyplot as plt
import numpy as np
from ipywidgets import Layout
from matplotlib.ticker import EngFormatter

from oscillodsp import dsp
from oscillodsp.oscillodsp_pb2 import TriggerMode, TriggerType
from oscillodsp.utils import Blinker, get_filename, modified_ylim

# Various definitions
DEFAULT_SAMPLE_QUANTIZE_BITS = 16
BLINK_LIST = ["Waiting", "Auto"]
DEFAULT_CH_COLOR = [
    "gold",
    "dodgerblue",
    "darkred",
    "darkblue",
    "black",
    "black",
    "black",
    "black",
]

# Define layouts
H_LAYOUT = Layout(width="150px")
V_LAYOUT = Layout(width="30px", height="4.1in")


def show_module_versions(modules):
    """
    Show Python and module versions
    """
    import importlib

    print("Python version:", sys.version)

    for m in modules:
        _ = importlib.import_module(m)
        print("{:s}: {:s}".format(m, _.__version__))


def observe(widget, trait_name):
    """
    This is required because may ipywidgets widgets don't accept decorator
    notation.

    Refer https://github.com/jupyter-widgets/ipywidgets/issues/717
    """

    def wrapper(func):
        widget.observe(func, trait_name)

    return wrapper


class Oscillo(widgets.HBox):
    """
    Main Oscillo class definition
    """

    def __init__(
        self,
        dsp_tty,
        dsp_bitrate,
        quantize_bits=DEFAULT_SAMPLE_QUANTIZE_BITS,
        loglevel=logging.WARNING,
        loghandler=logging.StreamHandler(sys.__stdout__),
        logformatter=logging.Formatter("oscillo.py: %(message)s"),
        dsp_loglevel=logging.WARNING,
    ):

        # Call widgets.HBox.__init__()
        # Refer https://kapernikov.com/ipywidgets-with-matplotlib/
        super().__init__()

        self.quantize_bits = quantize_bits

        # Set-up logger for debugging
        logger = logging.getLogger("oscillo")
        logger.setLevel(loglevel)
        loghandler.setFormatter(logformatter)
        logger.addHandler(loghandler)
        self.logger = logger

        # Identify underling OS (Linux, Windows, etc.) and determine asyncio
        # sleep length
        if sys.platform == "win32":
            self.sleep_len = 0.1
            self.logger.info(
                "Detected running on Windows, so using a special sleep length."
            )
        else:
            self.sleep_len = 0.01

        # Connecting to peer DSP
        self.peer = dsp.DSP(dsp_tty, dsp_bitrate, loglevel=dsp_loglevel)

        # Creating output object for Matplotlib figure
        output = widgets.Output()
        with output:
            self.fig, self.ax = plt.subplots(
                constrained_layout=True, figsize=(7, 4)
            )

        # Move the toolbar to the bottom
        self.fig.canvas.toolbar_position = "bottom"

        # Initialize other object variables which are related to plotting
        self.ch_active = 0  # channel currently paying attention by user
        self.trigmode = None
        self.trigtype = TriggerType.RisingEdge
        self.tscale = 0.0

        #
        # Define all widgets here
        #
        self.menu_ch = widgets.Dropdown(options=[], layout=H_LAYOUT)
        self.menu_trig = widgets.Dropdown(
            options=[
                ("Auto", TriggerMode.Auto),
                ("Normal", TriggerMode.Normal),
                ("Single", TriggerMode.Single),
            ],
            value=TriggerMode.Auto,
            layout=H_LAYOUT,
        )
        self.button_single = widgets.Button(description="Reset", disabled=True)
        self.text_trig_status = widgets.HTML("<b>Trigger Status</b>: Waiting")
        self.ftext_triglevel = widgets.FloatText(
            value=0.0, step=0.5, description="", layout=H_LAYOUT
        )
        self.button_stop = widgets.Button(
            description="Ready", button_style="primary"
        )
        self.button_save_image = widgets.Button(
            description="Save as Image File"
        )
        self.button_save_csv = widgets.Button(description="Save as CSV File")
        self.slider_mag = widgets.FloatSlider(
            value=1.0,
            min=0.1,
            max=1.91,
            step=0.1,
            orientation="vertical",
            layout=V_LAYOUT,
        )
        self.slider_ypos = widgets.FloatSlider(
            value=0.0,
            min=-1.0,
            max=1.0,
            orientation="vertical",
            layout=V_LAYOUT,
        )
        self.slider_tscale = widgets.FloatLogSlider(
            value=1 / 1e-3,  # reciprocal of time-scale
            base=10,
            min=0,
            max=9,
            step=1 / 3,
            orientation="horizontal",
            readout=False,
            layout=H_LAYOUT,
        )

        # Set Trigger Mode according to the menu selection
        self.trigmode_new = self.menu_trig.value  # for synchronous update
        self.triglevel_new = self.ftext_triglevel.value
        self.triglevel = 0.0

        #
        # Define callback functions respond GUI operations
        #
        @observe(self.menu_ch, "value")
        def _(change):
            self.ch_active = change.new
            self.slider_mag.value = self.mag[self.ch_active]
            self.slider_ypos.value = self.ypos[self.ch_active]

        @observe(self.menu_trig, "value")
        def _(change):
            self.trigmode_new = change.new

        @self.button_single.on_click
        def _(b):
            self.clear_trig = True

        @observe(self.ftext_triglevel, "value")
        def _(change):
            self.triglevel_new = change.new

        @self.button_stop.on_click
        def _(b):
            if self.stopped:
                for task in asyncio.all_tasks():
                    task.cancel()
                self.start()
            else:
                self.stopped = True
                self.button_stop.button_style = "danger"

        @self.button_save_image.on_click
        def _(b):
            s = get_filename(".png")
            plt.savefig(s)

        @self.button_save_csv.on_click
        def _(b):
            self.req_save_csv = True

        @observe(self.slider_mag, "value")
        def _(change):
            self.mag[self.ch_active] = change.new

        @observe(self.slider_ypos, "value")
        def _(change):
            self.ypos[self.ch_active] = change.new

        @observe(self.slider_tscale, "value")
        def _(change):
            self.tscale_new = 1 / change.new

        # Once call DSP to obtain various information need to set-up variables
        config_reply = self.peer.config(
            resolution=self.quantize_bits,
            trigmode=self.trigmode_new,
            trigtype=self.trigtype,
            ch_trig=0,
            triglevel=self.triglevel,
            timescale=0.0,
        )  # timescale as '0.0' means "Don't update timescale"

        # Set-up mag and ypos
        self.mag = []
        self.ypos = []
        for ch in config_reply.chconfig:
            self.mag.append(1.0)
            center = (ch.min + ch.max) / 2
            width = ch.max - ch.min
            self.ypos.append(-2.0 * center / width)

        # Create sub_ax (multiple ax on one subplot) of required numbers
        self.sub_ax = []
        for i in range(len(config_reply.chconfig) - 1):
            self.sub_ax.append(self.ax.twinx())

        # Configure Channel menu
        self.options = []
        self.view_enabled_ch = []
        for idx, ch in enumerate(config_reply.chconfig):
            self.options.append(("{:d}: {:s}".format(idx, ch.name), idx))
            self.view_enabled_ch.append(True)
        self.menu_ch.options = self.options

        # Configure default time-scale
        self.tscale_new = config_reply.default_timescale
        self.slider_tscale.min = math.log10(1 / config_reply.max_timescale)

        # Finally assemble all widgets in one box
        self.controls = widgets.HBox(
            [
                widgets.VBox(
                    [
                        widgets.VBox(
                            [
                                widgets.Box(layout=Layout(height="10px")),
                                widgets.HTML("Scale&nbsp;&nbsp;Pos"),
                            ],
                            layout=Layout(height="40px"),
                        ),
                        widgets.HBox(
                            [
                                self.slider_mag,
                                self.slider_ypos,
                            ]
                        ),
                    ]
                ),
                widgets.VBox(
                    [
                        widgets.HTML("<b>Active Channel</b>"),
                        self.menu_ch,
                        widgets.HTML("<b>Trigger Mode</b>"),
                        self.menu_trig,
                        self.button_single,
                        self.text_trig_status,
                        widgets.HTML("<b>Trigger Level</b>"),
                        self.ftext_triglevel,
                        widgets.HTML("<b>Horizontal Scale</b>"),
                        self.slider_tscale,
                        widgets.HTML("<hr>"),
                        self.button_stop,
                        self.button_save_image,
                        self.button_save_csv,
                    ]
                ),
            ]
        )
        self.children = [output, self.controls]

    # To allow GUI operations while repeating plot updates, we need to run
    # this function in background.  After experienced Python threading
    # functions didn't work correctly, finally adopted asyncio.
    #
    # Refer https://github.com/voila-dashboards/voila/issues/431
    async def oscillo_start(self):
        self.stopped = False
        self.clear_trig = False
        self.req_save_csv = False
        self.button_stop.description = "Run/Stop"
        self.button_stop.button_style = "success"

        blinker = Blinker()
        old_status = ""
        triggered = False
        last_ylim = None
        formatter = EngFormatter()

        while True:
            # When requested by UI, clear triggered flag synchronously
            if self.clear_trig:
                triggered = False
                self.clear_trig = False  # Reset the request

            # When requested by UI, clear triggered flag synchronously
            if (
                self.trigmode_new != self.trigmode
                or self.tscale_new != self.tscale
                or self.triglevel_new != self.triglevel
            ):
                self.trigmode = self.trigmode_new
                self.tscale = self.tscale_new
                self.triglevel = self.triglevel_new
                triggered = False
                self.button_single.disabled = (
                    self.trigmode != TriggerMode.Single
                )
                config_reply = self.peer.config(
                    resolution=self.quantize_bits,
                    trigmode=self.trigmode,
                    trigtype=self.trigtype,
                    ch_trig=0,
                    triglevel=self.triglevel,
                    timescale=self.tscale,
                )
                self.last_reply = config_reply

            # Once triggered for single-shot mode, re-use wave data
            if self.trigmode != TriggerMode.Single or not triggered:
                waves = self.peer.get_waves()  # Trying update wave data
                if len(waves.wave) > 0:
                    triggered = waves.triggered

            # Determine if plotting is required
            need_plot = len(waves.wave) > 0
            blank_screen = False

            # Update trigger status message
            if self.trigmode == TriggerMode.Auto:
                if triggered:
                    new_status = "Triggered"
                else:
                    new_status = "Auto"
            elif self.trigmode == TriggerMode.Normal:
                if need_plot:
                    new_status = "Triggered"
                else:
                    new_status = "Waiting"
            else:
                if triggered:
                    new_status = "Stopped"
                else:
                    new_status = "Waiting"
                    blank_screen = True

            # Blinking trigger status message
            if new_status != old_status:
                blinker.reset()
            if new_status in BLINK_LIST and not blinker.active():
                msg_status = ""
            else:
                msg_status = new_status
            self.text_trig_status.value = (
                "<b>Trigger Status</b>: " + msg_status
            )
            old_status = new_status

            if need_plot:
                # Generate CSV file header if required
                csv_samples = []
                if self.req_save_csv:
                    csv_printer = open(get_filename(".csv"), "w")
                    csv_printer.write("time [sec]")
                    for ch in self.last_reply.chconfig:
                        csv_printer.write(
                            ",{:s} [{:s}]".format(ch.name, ch.unit)
                        )
                    csv_printer.write("\n")

                # Now we can determine samples in a wave
                n_xsamples = len(waves.wave[0].samples)
                xlim = (-self.tscale / 2, self.tscale / 2)

                # ax.set_facecolor('#404040')
                # ax.axhline(y=0, color='k')
                # ax.axvline(x=0, color='k')

                chconfig_active = self.last_reply.chconfig[self.ch_active]

                if blank_screen:
                    self.ax.clear()
                    for ax in self.sub_ax:
                        ax.clear()
                        ax.set_axis_off()

                    self.ax.set_xlim(xlim)

                    if last_ylim is not None:
                        self.ax.set_ylim(last_ylim)
                else:
                    lines = []
                    # xser is common among multiple channels, so assign here
                    xser = np.linspace(xlim[0], xlim[1], n_xsamples)

                    ct_sub_ax = 0
                    for idx, wave in enumerate(waves.wave):
                        chconfig_idx = self.last_reply.chconfig[idx]

                        if idx == self.ch_active:
                            cur_ax = self.ax
                        else:
                            cur_ax = self.sub_ax[ct_sub_ax]
                            ct_sub_ax += 1

                        cur_ax.clear()

                        # Show only active channel axis
                        if idx != self.ch_active:
                            cur_ax.set_axis_off()

                        # Convert wave samples to original float values
                        ylim = (chconfig_idx.min, chconfig_idx.max)
                        yser = np.asarray(wave.samples, dtype=np.float32)
                        yser /= (1 << self.quantize_bits) / (ylim[1] - ylim[0])
                        yser += (ylim[0] + ylim[1]) / 2

                        # Keep yser samples for CSV output
                        if self.req_save_csv:
                            csv_samples.append(yser)

                        # Use a special ylim which taking account of mag and
                        # ypos, and set_ylim()
                        last_ylim = modified_ylim(
                            ylim, self.mag[idx], self.ypos[idx]
                        )  # save for blank_screen
                        cur_ax.set_ylim(last_ylim)

                        # Generating label for line plot
                        label = chconfig_idx.name
                        if idx == self.ch_active:
                            label += " (active)"

                        # Finally plot line of a channel
                        if self.view_enabled_ch[idx]:
                            (l,) = cur_ax.plot(
                                xser, yser, DEFAULT_CH_COLOR[idx], label=label
                            )

                            # Append the line to lines[] to show legend on the
                            # screen.
                            # Note that this is required ONLY for legend
                            # purpose.  Even without this "lines", plottings
                            # themselves are accomplished.
                            lines.append(l)

                # xlim is common among channels, so process here
                self.ax.set_xlim(xlim)
                self.ax.xaxis.set_major_formatter(formatter)
                self.ax.set_xlabel("[sec]")

                # ylim is configured for currently active channel
                self.ax.yaxis.set_major_formatter(formatter)
                self.ax.set_ylabel(
                    "{:s} [{:s}]".format(
                        chconfig_active.name, chconfig_active.unit
                    )
                )

                # Show legend here
                if len(lines) > 0:
                    self.ax.legend(
                        lines,
                        [_.get_label() for _ in lines],
                        loc="upper left",
                        # labelcolor='white',
                        facecolor="lightgray",
                        # edgecolor='white'
                    )

                # Show grid and finally draw everything
                self.ax.grid(True)
                self.fig.canvas.draw()

                # Save to CSV file if required
                if self.req_save_csv:
                    for i in range(n_xsamples):
                        csv_printer.write("{:e}".format(xser[i]))
                        for ch_id in range(len(self.last_reply.chconfig)):
                            csv_printer.write(
                                ",{:e}".format(csv_samples[ch_id][i])
                            )
                        csv_printer.write("\n")
                    csv_printer.close()
                    self.req_save_csv = False

            if self.stopped:
                break

            await asyncio.sleep(self.sleep_len)

    def start(self):
        """
        Start oscilloscope
        """
        asyncio.create_task(self.oscillo_start())

    def close(self):
        """
        Close oscilloscope
        """
        for task in asyncio.all_tasks():
            task.cancel()
        plt.close("all")
        self.peer.terminate()
        del self.peer
